File: src/main/java/uk/co/reecedunn/intellij/plugin/xqdoc/lexer/XQDocLexer.java
/*
 * Copyright (C) 2016 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.xqdoc.lexer;

import com.intellij.lexer.LexerBase;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import uk.co.reecedunn.intellij.plugin.core.lexer.CodePointRange;

import java.util.EmptyStackException;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class XQDocLexer extends LexerBase {
    private CodePointRange mTokenRange;
    private int mState;
    private final Stack<Integer> mStates = new Stack<>();
    private IElementType mType;

    public XQDocLexer() {
        mTokenRange = new CodePointRange();
    }

    // region States

    private void pushState(int state) {
        mStates.push(state);
    }

    private void popState() {
        try {
            mStates.pop();
        } catch (EmptyStackException e) {
            //
        }
    }

    private static final int STATE_DEFAULT = 0;
    private static final int STATE_CONTENTS = 1;
    private static final int STATE_TAGGED_CONTENTS = 2;
    private static final int STATE_ELEM_CONSTRUCTOR = 3;
    private static final int STATE_ELEM_CONTENTS = 4;
    private static final int STATE_ELEM_CONSTRUCTOR_CLOSING = 5;
    private static final int STATE_ATTRIBUTE_VALUE_QUOTE = 6;
    private static final int STATE_ATTRIBUTE_VALUE_APOS = 7;
    private static final int STATE_TRIM = 8;

    private void stateDefault() {
        int c = mTokenRange.getCodePoint();
        switch (c) {
            case CodePointRange.END_OF_BUFFER:
                mType = null;
                break;
            case '~':
                mTokenRange.match();
                mType = XQDocTokenType.XQDOC_COMMENT_MARKER;
                pushState(STATE_CONTENTS);
                pushState(STATE_TRIM);
                break;
            default:
                mTokenRange.seek(mTokenRange.getBufferEnd());
                mType = XQDocTokenType.COMMENT_CONTENTS;
                break;
        }
    }

    private void stateContents() {
        int c = mTokenRange.getCodePoint();
        switch (c) {
            case CodePointRange.END_OF_BUFFER:
                mType = null;
                break;
            case '<':
                mTokenRange.match();
                mType = XQDocTokenType.OPEN_XML_TAG;
                pushState(STATE_ELEM_CONSTRUCTOR);
                break;
            case '\n': // U+000A
            case '\r': // U+000D
                pushState(STATE_TRIM);
                stateTrim();
                break;
            default:
                while (true) switch (c) {
                    case '\n': // U+000A
                    case '\r': // U+000D
                        pushState(STATE_TRIM);
                        // fallthrough
                    case CodePointRange.END_OF_BUFFER:
                    case '<':
                        mType = XQDocTokenType.CONTENTS;
                        return;
                    default:
                        mTokenRange.match();
                        c = mTokenRange.getCodePoint();
                        break;
                }
        }
    }

    private void stateTaggedContents() {
        int c = mTokenRange.getCodePoint();
        if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
            while ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
                mTokenRange.match();
                c = mTokenRange.getCodePoint();
            }
            mType = sTagNames.getOrDefault(getTokenText(), XQDocTokenType.TAG);
        } else {
            popState();
            stateContents();
        }
    }

    private void stateElemConstructor(int state) {
        int c = mTokenRange.getCodePoint();
        switch (c) {
            case CodePointRange.END_OF_BUFFER:
                mType = null;
                break;
            case '0': case '1': case '2': case '3': case '4':
            case '5': case '6': case '7': case '8': case '9':
            case 'a': case 'b': case 'c': case 'd': case 'e': case 'f': case 'g':
            case 'h': case 'i': case 'j': case 'k': case 'l': case 'm': case 'n':
            case 'o': case 'p': case 'q': case 'r': case 's': case 't': case 'u':
            case 'v': case 'w': case 'x': case 'y': case 'z':
            case 'A': case 'B': case 'C': case 'D': case 'E': case 'F': case 'G':
            case 'H': case 'I': case 'J': case 'K': case 'L': case 'M': case 'N':
            case 'O': case 'P': case 'Q': case 'R': case 'S': case 'T': case 'U':
            case 'V': case 'W': case 'X': case 'Y': case 'Z':
                while ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) {
                    mTokenRange.match();
                    c = mTokenRange.getCodePoint();
                }
                mType = XQDocTokenType.XML_TAG;
                break;
            case ' ': case '\t': case '\r': case '\n':
                while (c == ' ' || c == '\t' || c == '\r' || c == '\n') {
                    mTokenRange.match();
                    c = mTokenRange.getCodePoint();
                }
                mType = XQDocTokenType.WHITE_SPACE;
                break;
            case '=':
                mTokenRange.match();
                mType = XQDocTokenType.XML_EQUAL;
                break;
            case '"':
                mTokenRange.match();
                mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_START;
                pushState(STATE_ATTRIBUTE_VALUE_QUOTE);
                break;
            case '\'':
                mTokenRange.match();
                mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_START;
                pushState(STATE_ATTRIBUTE_VALUE_APOS);
                break;
            case '/':
                mTokenRange.match();
                if (mTokenRange.getCodePoint() == '>') {
                    mTokenRange.match();
                    mType = XQDocTokenType.SELF_CLOSING_XML_TAG;
                    popState();
                } else {
                    mType = XQDocTokenType.INVALID;
                }
                break;
            case '>':
                mTokenRange.match();
                mType = XQDocTokenType.END_XML_TAG;
                popState();
                if (state == STATE_ELEM_CONSTRUCTOR) {
                    pushState(STATE_ELEM_CONTENTS);
                }
                break;
            default:
                mTokenRange.match();
                mType = XQDocTokenType.INVALID;
                break;
        }
    }

    private void stateElemContents() {
        int c = mTokenRange.getCodePoint();
        switch (c) {
            case CodePointRange.END_OF_BUFFER:
                mType = null;
                break;
            case '<':
                mTokenRange.match();
                if (mTokenRange.getCodePoint() == '/') {
                    mTokenRange.match();
                    mType = XQDocTokenType.CLOSE_XML_TAG;
                    popState();
                    pushState(STATE_ELEM_CONSTRUCTOR_CLOSING);
                } else {
                    mType = XQDocTokenType.OPEN_XML_TAG;
                    pushState(STATE_ELEM_CONSTRUCTOR);
                }
                break;
            default:
                mTokenRange.match();
                c = mTokenRange.getCodePoint();
                while (true) switch (c) {
                    case CodePointRange.END_OF_BUFFER:
                    case '<':
                        mType = XQDocTokenType.XML_ELEMENT_CONTENTS;
                        return;
                    default:
                        mTokenRange.match();
                        c = mTokenRange.getCodePoint();
                        break;
                }
        }
    }

    private void stateAttributeValue(int endChar) {
        int c = mTokenRange.getCodePoint();
        if (c == CodePointRange.END_OF_BUFFER) {
            mType = null;
        } else if (c == endChar) {
            mTokenRange.match();
            mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_END;
            popState();
        } else {
            while (c != CodePointRange.END_OF_BUFFER && c != endChar) {
                mTokenRange.match();
                c = mTokenRange.getCodePoint();
            }
            mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_CONTENTS;
        }
    }

    private void stateTrim() {
        int c = mTokenRange.getCodePoint();
        switch (c) {
            case CodePointRange.END_OF_BUFFER:
                mType = null;
                break;
            case ' ':
            case '\t':
                while (c == ' ' || c == '\t') {
                    mTokenRange.match();
                    c = mTokenRange.getCodePoint();
                }
                mType = XQDocTokenType.WHITE_SPACE;
                break;
            case '\r': // U+000D
            case '\n': // U+000A
                mTokenRange.match();
                if (c == '\r' && mTokenRange.getCodePoint() == '\n') {
                    mTokenRange.match();
                }

                c = mTokenRange.getCodePoint();
                while (c == ' ' || c == '\t') { // U+0020 || U+0009
                    mTokenRange.match();
                    c = mTokenRange.getCodePoint();
                }

                if (c == ':') {
                    mTokenRange.match();
                }

                mType = XQDocTokenType.TRIM;
                break;
            case '@':
                mTokenRange.match();
                mType = XQDocTokenType.TAG_MARKER;
                popState();
                pushState(STATE_TAGGED_CONTENTS);
                break;
            default:
                popState();
                advance();
                break;
        }
    }

    // endregion
    // region Lexer

    @Override
    public final void start(@NotNull CharSequence buffer, int startOffset, int endOffset, int initialState) {
        mTokenRange.start(buffer, startOffset, endOffset);
        pushState(initialState);
        advance();
    }

    @Override
    public final void advance() {
        mTokenRange.flush();
        try {
            mState = mStates.peek();
        } catch (EmptyStackException e) {
            mState = STATE_DEFAULT;
        }
        switch (mState) {
            case STATE_DEFAULT:
                stateDefault();
                break;
            case STATE_CONTENTS:
                stateContents();
                break;
            case STATE_TAGGED_CONTENTS:
                stateTaggedContents();
                break;
            case STATE_ELEM_CONSTRUCTOR:
            case STATE_ELEM_CONSTRUCTOR_CLOSING:
                stateElemConstructor(mState);
                break;
            case STATE_ELEM_CONTENTS:
                stateElemContents();
                break;
            case STATE_ATTRIBUTE_VALUE_QUOTE:
                stateAttributeValue('"');
                break;
            case STATE_ATTRIBUTE_VALUE_APOS:
                stateAttributeValue('\'');
                break;
            case STATE_TRIM:
                stateTrim();
                break;
            default:
                throw new AssertionError("Invalid state: " + mState);
        }
    }

    @Override
    public final int getState() {
        return mState;
    }

    @Override
    public final IElementType getTokenType() {
        return mType;
    }

    @Override
    public final int getTokenStart() {
        return mTokenRange.getStart();
    }

    @Override
    public final int getTokenEnd() {
        return mTokenRange.getEnd();
    }

    @Override
    @SuppressWarnings("NullableProblems") // jacoco Code Coverage reports an unchecked branch when @NotNull is used.
    public CharSequence getBufferSequence() {
        return mTokenRange.getBufferSequence();
    }

    @Override
    public final int getBufferEnd() {
        return mTokenRange.getBufferEnd();
    }

    // endregion
    // region Special Tag Names

    private static final Map<String, IElementType> sTagNames = new HashMap<>();

    static {
        sTagNames.put("author", XQDocTokenType.T_AUTHOR);
        sTagNames.put("deprecated", XQDocTokenType.T_DEPRECATED);
        sTagNames.put("error", XQDocTokenType.T_ERROR);
        sTagNames.put("param", XQDocTokenType.T_PARAM);
        sTagNames.put("return", XQDocTokenType.T_RETURN);
        sTagNames.put("see", XQDocTokenType.T_SEE);
        sTagNames.put("since", XQDocTokenType.T_SINCE);
        sTagNames.put("version", XQDocTokenType.T_VERSION);
    }

    // endregion
}

File: src/lang-xqdoc/main/uk/co/reecedunn/intellij/plugin/xqdoc/lexer/XQDocLexer.kt
/*
 * Copyright (C) 2016-2018 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.xqdoc.lexer

import uk.co.reecedunn.intellij.plugin.core.lexer.*

// region State Constants

private const val STATE_CONTENTS = 1
private const val STATE_TAGGED_CONTENTS = 2
private const val STATE_ELEM_CONSTRUCTOR = 3
private const val STATE_ELEM_CONTENTS = 4
private const val STATE_ELEM_CONSTRUCTOR_CLOSING = 5
private const val STATE_ATTRIBUTE_VALUE_QUOTE = 6
private const val STATE_ATTRIBUTE_VALUE_APOS = 7
private const val STATE_TRIM = 8
private const val STATE_PARAM_TAG_CONTENTS_START = 9
private const val STATE_PARAM_TAG_VARNAME = 10
private const val STATE_XQUERY_CONTENTS = 11
private const val STATE_XQUERY_CONTENTS_TRIM = 12

// endregion
// region Special Tag Names

private val TAG_NAMES = mapOf(
    "author" to XQDocTokenType.T_AUTHOR,
    "deprecated" to XQDocTokenType.T_DEPRECATED,
    "error" to XQDocTokenType.T_ERROR,
    "param" to XQDocTokenType.T_PARAM,
    "return" to XQDocTokenType.T_RETURN,
    "see" to XQDocTokenType.T_SEE,
    "since" to XQDocTokenType.T_SINCE,
    "version" to XQDocTokenType.T_VERSION
)

// endregion

class XQDocLexer : LexerImpl(STATE_CONTENTS, CodePointRangeImpl()) {
    // region States

    private fun matchEntityReference() {
        mTokenRange.match()
        var cc = CharacterClass.getCharClass(mTokenRange.codePoint)
        if (cc == CharacterClass.NAME_START_CHAR) {
            mTokenRange.match()
            cc = CharacterClass.getCharClass(mTokenRange.codePoint)
            while (cc == CharacterClass.NAME_START_CHAR || cc == CharacterClass.DIGIT) {
                mTokenRange.match()
                cc = CharacterClass.getCharClass(mTokenRange.codePoint)
            }
            mType = if (cc == CharacterClass.SEMICOLON) {
                mTokenRange.match()
                XQDocTokenType.PREDEFINED_ENTITY_REFERENCE
            } else {
                XQDocTokenType.PARTIAL_ENTITY_REFERENCE
            }
        } else if (cc == CharacterClass.HASH) {
            mTokenRange.match()
            var c = mTokenRange.codePoint
            if (c == 'x'.toInt()) {
                mTokenRange.match()
                c = mTokenRange.codePoint
                if (c >= '0'.toInt() && c <= '9'.toInt() || c >= 'a'.toInt() && c <= 'f'.toInt() || c >= 'A'.toInt() && c <= 'F'.toInt()) {
                    while (c >= '0'.toInt() && c <= '9'.toInt() || c >= 'a'.toInt() && c <= 'f'.toInt() || c >= 'A'.toInt() && c <= 'F'.toInt()) {
                        mTokenRange.match()
                        c = mTokenRange.codePoint
                    }
                    mType = if (c == ';'.toInt()) {
                        mTokenRange.match()
                        XQDocTokenType.CHARACTER_REFERENCE
                    } else {
                        XQDocTokenType.PARTIAL_ENTITY_REFERENCE
                    }
                } else if (c == ';'.toInt()) {
                    mTokenRange.match()
                    mType = XQDocTokenType.EMPTY_ENTITY_REFERENCE
                } else {
                    mType = XQDocTokenType.PARTIAL_ENTITY_REFERENCE
                }
            } else if (c >= '0'.toInt() && c <= '9'.toInt()) {
                mTokenRange.match()
                while (c >= '0'.toInt() && c <= '9'.toInt()) {
                    mTokenRange.match()
                    c = mTokenRange.codePoint
                }
                mType = if (c == ';'.toInt()) {
                    mTokenRange.match()
                    XQDocTokenType.CHARACTER_REFERENCE
                } else {
                    XQDocTokenType.PARTIAL_ENTITY_REFERENCE
                }
            } else if (c == ';'.toInt()) {
                mTokenRange.match()
                mType = XQDocTokenType.EMPTY_ENTITY_REFERENCE
            } else {
                mType = XQDocTokenType.PARTIAL_ENTITY_REFERENCE
            }
        } else if (cc == CharacterClass.SEMICOLON) {
            mTokenRange.match()
            mType = XQDocTokenType.EMPTY_ENTITY_REFERENCE
        } else {
            mType = XQDocTokenType.PARTIAL_ENTITY_REFERENCE
        }
    }

    private fun stateDefault() {
        val c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            '~'.toInt() -> {
                mTokenRange.match()
                mType = XQDocTokenType.XQDOC_COMMENT_MARKER
                pushState(STATE_CONTENTS)
                pushState(STATE_TRIM)
            }
            else -> {
                pushState(STATE_XQUERY_CONTENTS)
                pushState(STATE_TRIM)
                advance()
            }
        }
    }

    private fun stateContents() {
        var c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            '<'.toInt() -> {
                mTokenRange.match()
                mType = XQDocTokenType.OPEN_XML_TAG
                pushState(STATE_ELEM_CONSTRUCTOR)
            }
            '\n'.toInt(), '\r'.toInt() -> { // U+000A, U+000D
                pushState(STATE_TRIM)
                stateTrim(STATE_TRIM)
            }
            '&'.toInt() -> matchEntityReference() // XML PredefinedEntityRef and CharRef
            else -> while (true)
                when (c) {
                    '\n'.toInt(), '\r'.toInt() -> { // U+000A, U+000D
                        pushState(STATE_TRIM)
                        mType = XQDocTokenType.CONTENTS
                        return
                    }
                    CodePointRange.END_OF_BUFFER, '<'.toInt(), '&'.toInt() -> {
                        mType = XQDocTokenType.CONTENTS
                        return
                    }
                    else -> {
                        mTokenRange.match()
                        c = mTokenRange.codePoint
                    }
                }
        }
    }

    private fun stateXQueryContents() {
        var c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            '\n'.toInt(), '\r'.toInt() -> { // U+000A, U+000D
                pushState(STATE_XQUERY_CONTENTS_TRIM)
                stateTrim(STATE_XQUERY_CONTENTS_TRIM)
            }
            else -> while (true)
                when (c) {
                    '\n'.toInt(), '\r'.toInt() -> { // U+000A, U+000D
                        pushState(STATE_XQUERY_CONTENTS_TRIM)
                        mType = XQDocTokenType.CONTENTS
                        return
                    }
                    CodePointRange.END_OF_BUFFER -> {
                        mType = XQDocTokenType.CONTENTS
                        return
                    }
                    else -> {
                        mTokenRange.match()
                        c = mTokenRange.codePoint
                    }
                }
        }
    }

    private fun stateTaggedContents() {
        var c = mTokenRange.codePoint
        if (c >= 'a'.toInt() && c <= 'z'.toInt() || c >= 'A'.toInt() && c <= 'Z'.toInt() || c >= '0'.toInt() && c <= '9'.toInt()) {
            while (c >= 'a'.toInt() && c <= 'z'.toInt() || c >= 'A'.toInt() && c <= 'Z'.toInt() || c >= '0'.toInt() && c <= '9'.toInt()) {
                mTokenRange.match()
                c = mTokenRange.codePoint
            }
            mType = TAG_NAMES[tokenText] ?: XQDocTokenType.TAG
            if (mType === XQDocTokenType.T_PARAM) {
                popState()
                pushState(STATE_PARAM_TAG_CONTENTS_START)
            }
        } else if (c == ' '.toInt() || c == '\t'.toInt()) {
            while (c == ' '.toInt() || c == '\t'.toInt()) {
                mTokenRange.match()
                c = mTokenRange.codePoint
            }
            mType = XQDocTokenType.WHITE_SPACE
            popState()
        } else {
            popState()
            stateContents()
        }
    }

    private fun stateParamTagContentsStart() {
        var c = mTokenRange.codePoint
        if (c == ' '.toInt() || c == '\t'.toInt()) {
            while (c == ' '.toInt() || c == '\t'.toInt()) {
                mTokenRange.match()
                c = mTokenRange.codePoint
            }
            mType = XQDocTokenType.WHITE_SPACE
        } else if (c == '$'.toInt()) {
            mTokenRange.match()
            mType = XQDocTokenType.VARIABLE_INDICATOR
            popState()
            pushState(STATE_PARAM_TAG_VARNAME)
        } else {
            popState()
            stateContents()
        }
    }

    private fun stateParamTagVarName() {
        var cc = CharacterClass.getCharClass(mTokenRange.codePoint)
        when (cc) {
            CharacterClass.WHITESPACE -> {
                mTokenRange.match()
                while (CharacterClass.getCharClass(mTokenRange.codePoint) == CharacterClass.WHITESPACE)
                    mTokenRange.match()
                mType = XQDocTokenType.WHITE_SPACE
                popState()
            }
            CharacterClass.NAME_START_CHAR // XQuery/XML NCName token rules.
            -> {
                mTokenRange.match()
                cc = CharacterClass.getCharClass(mTokenRange.codePoint)
                while (
                    cc == CharacterClass.NAME_START_CHAR ||
                    cc == CharacterClass.DIGIT ||
                    cc == CharacterClass.DOT ||
                    cc == CharacterClass.HYPHEN_MINUS ||
                    cc == CharacterClass.NAME_CHAR
                ) {
                    mTokenRange.match()
                    cc = CharacterClass.getCharClass(mTokenRange.codePoint)
                }
                mType = XQDocTokenType.NCNAME
            }
            else -> {
                popState()
                stateContents()
            }
        }
    }

    private fun stateElemConstructor(state: Int) {
        var c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            '0'.toInt(), '1'.toInt(), '2'.toInt(), '3'.toInt(), '4'.toInt(), '5'.toInt(), '6'.toInt(), '7'.toInt(),
            '8'.toInt(), '9'.toInt(), 'a'.toInt(), 'b'.toInt(), 'c'.toInt(), 'd'.toInt(), 'e'.toInt(), 'f'.toInt(),
            'g'.toInt(), 'h'.toInt(), 'i'.toInt(), 'j'.toInt(), 'k'.toInt(), 'l'.toInt(), 'm'.toInt(), 'n'.toInt(),
            'o'.toInt(), 'p'.toInt(), 'q'.toInt(), 'r'.toInt(), 's'.toInt(), 't'.toInt(), 'u'.toInt(), 'v'.toInt(),
            'w'.toInt(), 'x'.toInt(), 'y'.toInt(), 'z'.toInt(), 'A'.toInt(), 'B'.toInt(), 'C'.toInt(), 'D'.toInt(),
            'E'.toInt(), 'F'.toInt(), 'G'.toInt(), 'H'.toInt(), 'I'.toInt(), 'J'.toInt(), 'K'.toInt(), 'L'.toInt(),
            'M'.toInt(), 'N'.toInt(), 'O'.toInt(), 'P'.toInt(), 'Q'.toInt(), 'R'.toInt(), 'S'.toInt(), 'T'.toInt(),
            'U'.toInt(), 'V'.toInt(), 'W'.toInt(), 'X'.toInt(), 'Y'.toInt(), 'Z'.toInt() -> {
                while (c >= 'a'.toInt() && c <= 'z'.toInt() || c >= 'A'.toInt() && c <= 'Z'.toInt() || c >= '0'.toInt() && c <= '9'.toInt()) {
                    mTokenRange.match()
                    c = mTokenRange.codePoint
                }
                mType = XQDocTokenType.XML_TAG
            }
            ' '.toInt(), '\t'.toInt(), '\r'.toInt(), '\n'.toInt() -> {
                while (c == ' '.toInt() || c == '\t'.toInt() || c == '\r'.toInt() || c == '\n'.toInt()) {
                    mTokenRange.match()
                    c = mTokenRange.codePoint
                }
                mType = XQDocTokenType.WHITE_SPACE
            }
            '='.toInt() -> {
                mTokenRange.match()
                mType = XQDocTokenType.XML_EQUAL
            }
            '"'.toInt() -> {
                mTokenRange.match()
                mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_START
                pushState(STATE_ATTRIBUTE_VALUE_QUOTE)
            }
            '\''.toInt() -> {
                mTokenRange.match()
                mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_START
                pushState(STATE_ATTRIBUTE_VALUE_APOS)
            }
            '/'.toInt() -> {
                mTokenRange.match()
                if (mTokenRange.codePoint == '>'.toInt()) {
                    mTokenRange.match()
                    mType = XQDocTokenType.SELF_CLOSING_XML_TAG
                    popState()
                } else {
                    mType = XQDocTokenType.INVALID
                }
            }
            '>'.toInt() -> {
                mTokenRange.match()
                mType = XQDocTokenType.END_XML_TAG
                popState()
                if (state == STATE_ELEM_CONSTRUCTOR) {
                    pushState(STATE_ELEM_CONTENTS)
                }
            }
            else -> {
                mTokenRange.match()
                mType = XQDocTokenType.INVALID
            }
        }
    }

    private fun stateElemContents() {
        var c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            '<'.toInt() -> {
                mTokenRange.match()
                if (mTokenRange.codePoint == '/'.toInt()) {
                    mTokenRange.match()
                    mType = XQDocTokenType.CLOSE_XML_TAG
                    popState()
                    pushState(STATE_ELEM_CONSTRUCTOR_CLOSING)
                } else {
                    mType = XQDocTokenType.OPEN_XML_TAG
                    pushState(STATE_ELEM_CONSTRUCTOR)
                }
            }
            '&'.toInt() -> matchEntityReference() // XML PredefinedEntityRef and CharRef
            else -> {
                mTokenRange.match()
                c = mTokenRange.codePoint
                while (true)
                    when (c) {
                        CodePointRange.END_OF_BUFFER, '<'.toInt(), '&'.toInt() -> {
                            mType = XQDocTokenType.XML_ELEMENT_CONTENTS
                            return
                        }
                        else -> {
                            mTokenRange.match()
                            c = mTokenRange.codePoint
                        }
                    }
            }
        }
    }

    private fun stateAttributeValue(endChar: Int) {
        var c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            endChar -> {
                mTokenRange.match()
                mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_END
                popState()
            }
            else -> {
                while (c != CodePointRange.END_OF_BUFFER && c != endChar) {
                    mTokenRange.match()
                    c = mTokenRange.codePoint
                }
                mType = XQDocTokenType.XML_ATTRIBUTE_VALUE_CONTENTS
            }
        }
    }

    private fun stateTrim(state: Int) {
        var c = mTokenRange.codePoint
        when (c) {
            CodePointRange.END_OF_BUFFER -> mType = null
            ' '.toInt(), '\t'.toInt() -> {
                while (c == ' '.toInt() || c == '\t'.toInt()) {
                    mTokenRange.match()
                    c = mTokenRange.codePoint
                }
                mType = XQDocTokenType.WHITE_SPACE
            }
            '\r'.toInt(), '\n'.toInt() -> { // U+000D, U+000A
                mTokenRange.match()
                if (c == '\r'.toInt() && mTokenRange.codePoint == '\n'.toInt()) {
                    mTokenRange.match()
                }

                c = mTokenRange.codePoint
                while (c == ' '.toInt() || c == '\t'.toInt()) { // U+0020 || U+0009
                    mTokenRange.match()
                    c = mTokenRange.codePoint
                }

                if (c == ':'.toInt()) {
                    mTokenRange.match()
                }

                mType = XQDocTokenType.TRIM
            }
            '@'.toInt() -> {
                if (state == STATE_TRIM) {
                    mTokenRange.match()
                    mType = XQDocTokenType.TAG_MARKER
                    popState()
                    pushState(STATE_TAGGED_CONTENTS)
                } else {
                    popState()
                    advance()
                }
            }
            else -> {
                popState()
                advance()
            }
        }
    }

    // endregion
    // region Lexer

    override fun advance() {
        when (nextState()) {
            STATE_DEFAULT -> stateDefault()
            STATE_CONTENTS -> stateContents()
            STATE_TAGGED_CONTENTS -> stateTaggedContents()
            STATE_ELEM_CONSTRUCTOR, STATE_ELEM_CONSTRUCTOR_CLOSING -> stateElemConstructor(state)
            STATE_ELEM_CONTENTS -> stateElemContents()
            STATE_ATTRIBUTE_VALUE_QUOTE -> stateAttributeValue('"'.toInt())
            STATE_ATTRIBUTE_VALUE_APOS -> stateAttributeValue('\''.toInt())
            STATE_TRIM, STATE_XQUERY_CONTENTS_TRIM -> stateTrim(state)
            STATE_PARAM_TAG_CONTENTS_START -> stateParamTagContentsStart()
            STATE_PARAM_TAG_VARNAME -> stateParamTagVarName()
            STATE_XQUERY_CONTENTS -> stateXQueryContents()
            else -> throw AssertionError("Invalid state: $state")
        }
    }

    // endregion
}
